### PR TITLE
chore: configure codeql vulnerability scanning

### PR DESCRIPTION
**Pull request by kenneth-rosario**
Configure Github's CodeQL vulnerability scanning and fix findings.

### CODE REVIEW

File: funcframework/events.go [hunk 1]
```diff
@@ -126,7 +126,9 @@ func getBackgroundEvent(body []byte, path string) (*metadata.Metadata, interface
 	if possible.BackgroundEvent == nil && possible.LegacyPushSubscriptionEvent != nil {
 		topic, err := pubsub.ExtractTopicFromRequestPath(path)
 		if err != nil {
-			fmt.Printf("WARNING: %s", err)
+			escapedLog := strings.ReplaceAll(fmt.Sprintf("WARNING: %s", err), "\n", "")
+			escapedLog = strings.ReplaceAll(escapedLog, "\r", "")
+			fmt.Println(escapedLog)
```
[the remark this answers is not in the document]
Comment (HKWinterhalter) [reply funcframework/events.go:131]: also, what's the reason for removing newlines?

File: funcframework/framework.go [hunk 2]
```diff
@@ -293,12 +294,21 @@ func fmtFunctionError(err interface{}) string {
 	return formatted
 }
 
+func sanitizeMessage(message string) string {
+	newMessage := strings.ReplaceAll(message, "\n", "")
+	newMessage = strings.ReplaceAll(newMessage, "\r", "")
+	newMessage = html.EscapeString(newMessage)
+	return newMessage
```
Comment (HKWinterhalter) [inline funcframework/framework.go:301]: same comment as above, consolidate this to one or two lines

**Review by HKWinterhalter — APPROVED** [review]

File: funcframework/events.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kenneth-rosario) [reply funcframework/events.go]: Synced offline

File: .github/workflows/codeql.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "master" ]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "master" ]
+  schedule:
```
Comment (anniefu) [inline .github/workflows/codeql.yml:9]: [nit] do you need to include this section if it's not running?

File: .github/workflows/codeql.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    permissions:
+      actions: read
+      contents: read
+      security-events: write
+
+    strategy:
+      fail-fast: false
+      matrix:
+        language: [ 'go' ]
+        # CodeQL supports [ 'cpp', 'csharp', 'go', 'java', 'javascript', 'python', 'ruby' ]
```
Comment (anniefu) [inline .github/workflows/codeql.yml:25]: [nit] can probably remove these comments except maybe the last line that links to docs. probably same below

File: .github/workflows/codeql.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v3
+
+    # Initializes the CodeQL tools for scanning.
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v2
+      with:
+        languages: ${{ matrix.language }}
```
Comment (anniefu) [inline .github/workflows/codeql.yml:38]: [nit] Given the repo we're in, we can probably just fill this in with Go and not use a matrix

File: funcframework/events.go [hunk 6]
```diff
@@ -126,7 +126,9 @@ func getBackgroundEvent(body []byte, path string) (*metadata.Metadata, interface
 	if possible.BackgroundEvent == nil && possible.LegacyPushSubscriptionEvent != nil {
 		topic, err := pubsub.ExtractTopicFromRequestPath(path)
 		if err != nil {
-			fmt.Printf("WARNING: %s", err)
+			replacer := strings.NewReplacer("\n", "", "\r", "")
```
Comment (anniefu) [inline funcframework/events.go:129]: Based on the alert's stack trace, we should probably do this on line 53 of pubsub.go and apply it just to the path.

We probably don't want to remove whitespaces everywhere in the error message as something in our own code might have added it to make the error message more readable.

File: funcframework/framework.go [hunk 7]
```diff
@@ -293,12 +294,15 @@ func fmtFunctionError(err interface{}) string {
...
+}
+
 func writeHTTPErrorResponse(w http.ResponseWriter, statusCode int, status, msg string) {
-	// Ensure logs end with a newline otherwise they are grouped incorrectly in SD.
-	if !strings.HasSuffix(msg, "\n") {
-		msg += "\n"
-	}
-	fmt.Fprint(os.Stderr, msg)
+	// Sanitize message before logging and writing to response
+	escapedMsg := sanitizeMessage(msg)
```
Comment (anniefu) [inline funcframework/framework.go:304]: We actually need to keep the old behavior here of *adding* a newline to the end of the message. Cloud Logging uses newlines to detect when a new log starts. So if a user didn't add a newline to the end of their function's log like this:

```
log.Print("this doesn't end with a new line")
```

And you invoke the function twice in a row, their Cloud Logging might show something like

`this doesn't end with a new linethis doesn't end with a new line`

When you really want a separate line for each invocation. (Or something similar to that, I dont' remember the exact scenario).

**Review by anniefu — CHANGES_REQUESTED** [review]

File: internal/events/pubsub/pubsub.go [hunk 8]
```diff
@@ -50,8 +51,10 @@ func ExtractTopicFromRequestPath(path string) (string, error) {
 	re := regexp.MustCompile(`(projects\/[^/?]+\/topics\/[^/?]+)/*`)
 	matches := re.FindStringSubmatch(path)
 	if matches == nil {
+		replacer := strings.NewReplacer("\n", "", "\r", "")
```
Comment (anniefu) [inline internal/events/pubsub/pubsub.go:54]: [nit] add comment for why we're doing this

File: funcframework/framework.go [hunk 9]
```diff
@@ -309,5 +310,5 @@ func writeHTTPErrorResponse(w http.ResponseWriter, statusCode int, status, msg s
 
 	w.Header().Set(functionStatusHeader, status)
 	w.WriteHeader(statusCode)
-	fmt.Fprint(w, msg)
+	fmt.Fprint(w, html.EscapeString(msg))
```
Comment (anniefu) [inline funcframework/framework.go:313]: Was the code scanning complaining about this one? If not, we probably want to leave it alone as there may be customers expecting their strings to be formatted a certain way and we're just changing it underneath them.

**Review by anniefu — APPROVED** [review]
Similar to PHP one, use `chore:` instead of `feat:` so we don't impact the version number of the next release

**Pull request merged**